lib/mcrypt.php

```diff
@@ -180,7 +180,7 @@ function phpseclib_set_key(Base $td, $key)
      */
     function phpseclib_set_iv(Base $td, $iv)
     {
-        if ($td->mcrypt_mode != 'stream' && $td->mcrypt_mode != 'ecb') {
```

Comment:
The old code checked to see if `$td->mcrypt_mode` was equal to `'stream'`. The new code does not. This is an issue because, in phpseclib:dev-master, setting an IV when one isn't required results in an exception being thrown. phpseclib:dev-master is a lot more strict than phpseclib:~2.0. The reason being kinda motivated by sarciszewski, the guy who spear headed the push to kill mcrypt.

In particular, sarciszewski's has heavily promoted libsodium because it makes it so you can't mess with the cryptographic primitives yourself. libsodium manages the cryptographic primitives for you so you can't make mistakes using them yourself. Kinda like why you'd want to use prepared statements in MySQL. But more than that, he's also been critical of other libraries (eg. OpenSSL) for enabling bad practices. phpseclib needs these cryptographic primitives to implement OpenSSL and to process X.509 certs, etc, but to discourage their use and to stave off some criticism phpseclib:dev-master is a lot less error tolerant.

Anyway, let's say you have the following bit of code:

```php
$td = mcrypt_module_open(MCRYPT_ARCFOUR, '', MCRYPT_MODE_STREAM, '');
mcrypt_generic_init($td, str_repeat('z', 16), 'x');
```

With native mcrypt you'll just get this:

```
Warning: mcrypt_generic_init(): Iv size incorrect; supplied length: 16, needed: 0 in /path/to/file.php
```

But with these changes to mcrypt_compat you'll get this:

```
Warning: mcrypt_generic_init(): Iv size incorrect; supplied length: 16, needed: 0 in /path/to/file.php

Fatal error: Uncaught InvalidArgumentException: This algorithm does not use an IV. in /path/to/vendor/phpseclib/ph
pseclib/phpseclib/Crypt/Common/SymmetricKey.php:542
Stack trace:
#0 /path/to/vendor/phpseclib/mcrypt_compat/lib/mcrypt.php(186): phpseclib\Crypt\Common\SymmetricKey->setIV('')
#1 /path/to/vendor/phpseclib/mcrypt_compat/lib/mcrypt.php(662): phpseclib_set_iv(Object(phpseclib\Crypt\RC4), '')
#2 /path/to/file.php(xx): phpseclib_mcrypt_generic_init(Object(phpseclib\Crypt\RC4), 'zzzzzzzzzzzzzzz...', 'zzzzzz
zzzzzzzzz...')
#3 {main}
  thrown in /path/to/vendor/phpseclib/phpseclib/phpseclib/Crypt/Common/SymmetricKey.php on line 542
```

I'd add a unit test to make sure that that Fatal error isn't thrown but I'm not actually sure how to do that with PHPUnit since PHPUnit turns the E_USER_WARNING into an Exception. ie. The InvalidArgumentException isn't thrown because PHPUnit's error handler throws a PHPUnit_Framework_Error_Warning exception when it encounters the E_USER_WARNING.

Reply:
Hey there, thanks for the detailed writeup. The intended behaviour for `stream` was something I was not sure about. Judging from the comment on `phpseclib_mcrypt_enc_get_iv_size` I somehow derived that there were some configurations where an IV could be used in stream mode:

    An IV is used in cbc, cfb and ofb modes, and in some algorithms
     * in stream mode.

I have now added `stream` to the "without-IV" list and will push the branch again in a few minutes.

However, there is now some quite strange behaviour - mcrypt seems to emit this notice inconsistently. It's not emitted for ECB, but it is for streams. Check the failing tests on this branch to see what I mean. Now I could make this go away with some tailored `if`s, but there might be some rationale behind it that I'm not aware of...

Reply:
> An IV is used in cbc, cfb and ofb modes, and in some algorithms in stream mode.

It depends on the stream cipher. RC4 doesn't support IV's but other stream ciphers might. But currently phpseclib doesn't support any stream ciphers other than RC4. If phpseclib eventually does get support for stream ciphers that do accept an IV mcrypt_compat can be updated accordingly but until then I think just checking to see if it's a stream cipher is sufficient.

> However, there is now some quite strange behaviour - mcrypt seems to emit this notice inconsistently. It's not emitted for ECB, but it is for streams. Check the failing tests on this branch to see what I mean.

I'll check it out as time permits.

Thanks!

Reply:
One of the changes you made to phpseclib_mcrypt_generic_init is to do this:

```php
        $iv_size = phpseclib_mcrypt_enc_get_iv_size($td);
        if (!phpseclib_mcrypt_module_is_iv_mode($td->mcrypt_mode)) {
            $iv_size = 0;
        }
```

mcrypt_generic_init() doesn't do that. mcrypt_enc_get_iv_size() returns 8 as does phpseclib_mcrypt_enc_get_iv_size() but then you're just kinda overwriting it.

With streams that don't use an IV (eg. arcfour), both mcrypt_enc_get_iv_size() and phpseclib_mcrypt_enc_get_iv_size() return 0 and thus an IV of size 0 is a-ok.

Sample code demonstrating this:

https://3v4l.org/lMp90

So I'd say just remove this bit and you'll be golden:

```php
        if (!phpseclib_mcrypt_module_is_iv_mode($td->mcrypt_mode)) {
            $iv_size = 0;
        }
```